pkg/apis/duck/v1beta1/channelable_types.go

```diff
@@ -0,0 +1,154 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	eventingduckv1alpha1 "knative.dev/eventing/pkg/apis/duck/v1alpha1"
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+)
+
+// +genduck
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Channelable is a skeleton type wrapping Subscribable and Addressable in the manner we expect resource writers
+// defining compatible resources to embed it. We will typically use this type to deserialize
+// Channelable ObjectReferences and access their subscription and address data.  This is not a real resource.
+type Channelable struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the part where the Channelable fulfills the Subscribable contract.
+	Spec ChannelableSpec `json:"spec,omitempty"`
+
+	Status ChannelableStatus `json:"status,omitempty"`
+}
+
+// ChannelableSpec contains Spec of the Channelable object
+type ChannelableSpec struct {
+	SubscribableTypeSpec `json:",inline"`
+
+	// DeliverySpec contains options controlling the event delivery
+	// +optional
+	Delivery *eventingduckv1alpha1.DeliverySpec `json:"delivery,omitempty"`
+}
+
+// ChannelableStatus contains the Status of a Channelable object.
+type ChannelableStatus struct {
+	// inherits duck/v1 Status, which currently provides:
+	// * ObservedGeneration - the 'Generation' of the Service that was last processed by the controller.
+	// * Conditions - the latest available observations of a resource's current state.
+	duckv1.Status `json:",inline"`
+	// AddressStatus is the part where the Channelable fulfills the Addressable contract.
+	duckv1.AddressStatus `json:",inline"`
+	// Subscribers is populated with the statuses of each of the Channelable's subscribers.
+	SubscribableTypeStatus `json:",inline"`
+	// ErrorChannel is set by the channel when it supports native error handling via a channel
+	// +optional
+	ErrorChannel *corev1.ObjectReference `json:"errorChannel,omitempty"`
```

Comment:
when you say ref, you mean pointer? Or something else? If a Channel does not support (and none are provided via spec, it could be missing)? And furthermore, are you saying this should be a apis.URL?


Fine with the name change.

Reply:
Yeah, I think Knative does object refs in specs, and resolved uris in status. BUT it depends on if this extra channel is being created for the channel and not given to the channel, then this might make sense.

So if the goal here is for the parent channel to advertise that it has made another channel for you to subscribe to errors on, then an object ref makes sense and I would just call it a `DeadLetterChannel` and keep it an object ref.

Reply:
`DeadLetterChannel` for messages that failed to be delivered. 

In the future I could also see `ErrorChannel` or `ErrorSink` as general errors that the channel would like to emit but those might not be because of incoming messages (like they are messages generated from the control plane).

Reply:
ack :)